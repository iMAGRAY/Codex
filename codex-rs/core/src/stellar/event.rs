use crate::resilience::ConflictId;
use crate::resilience::ResolutionState;
use crate::stellar::action::StellarAction;
use crate::stellar::persona::StellarPersona;
use serde::Deserialize;
use serde::Serialize;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "kebab-case")]
pub enum KernelEvent {
    Info {
        message: String,
    },
    Submission {
        text: String,
    },
    CacheStored {
        key: String,
    },
    ConflictResolution {
        conflict_id: ConflictId,
        state: ResolutionState,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StellarCliEvent {
    pub persona: StellarPersona,
    pub action: StellarAction,
}

impl StellarCliEvent {
    pub fn new(persona: StellarPersona, action: StellarAction) -> Self {
        Self { persona, action }
    }
}
