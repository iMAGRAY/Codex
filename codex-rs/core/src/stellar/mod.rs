//! Stellar core kernel primitives shared between the TUI and CLI front-ends.
//!
//! The kernel is intentionally platform-agnostic: it expresses user intent as
//! [`StellarAction`] values scoped by [`StellarPersona`] profiles, validates
//! them via [`InputGuard`], and applies updates to the in-memory
//! [`StellarKernel`] state machine. UI layers (TUI, CLI bridge, automated
//! workflows) consume the resulting [`KernelSnapshot`] to render layouts or
//! emit structured events.

mod action;
mod event;
mod guard;
mod persona;
mod rbac;
mod snapshot;
mod state;

pub use action::StellarAction;
pub use action::StellarActionId;
pub use event::KernelEvent;
pub use event::StellarCliEvent;
pub use guard::GuardError;
pub use guard::InputGuard;
pub use persona::StellarPersona;
pub use rbac::is_action_allowed;
pub use snapshot::ConfidenceSnapshot;
pub use snapshot::GoldenPathHint;
pub use snapshot::KernelSnapshot;
pub use snapshot::LayoutMode;
pub use snapshot::PaneFocus;
pub use snapshot::RiskAlert;
pub use snapshot::RiskSeverity;
pub use snapshot::RunbookShortcut;
pub use state::ActionApplied;
pub use state::StellarKernel;

#[cfg(test)]
mod tests;
