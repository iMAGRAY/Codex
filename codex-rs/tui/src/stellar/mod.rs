mod controller;
mod keymap;
mod view;

pub use controller::ControllerOutcome;
pub use controller::StellarController;
pub use view::StellarView;

#[cfg(test)]
mod tests;
