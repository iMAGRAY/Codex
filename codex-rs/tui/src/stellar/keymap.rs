use codex_core::stellar::StellarAction;
use codex_core::stellar::StellarPersona;
use crossterm::event::KeyCode;
use crossterm::event::KeyEvent;
use crossterm::event::KeyEventKind;
use crossterm::event::KeyModifiers;
use std::collections::HashMap;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
struct KeyChord {
    code: KeyCode,
    modifiers: KeyModifiers,
}

impl KeyChord {
    fn new(mut code: KeyCode, modifiers: KeyModifiers) -> Self {
        if let KeyCode::Char(c) = code {
            code = KeyCode::Char(c.to_ascii_lowercase());
        }
        Self { code, modifiers }
    }
}

#[derive(Debug, Default)]
pub struct KeymapEngine {
    global: HashMap<KeyChord, StellarAction>,
    persona_overlays: HashMap<StellarPersona, HashMap<KeyChord, StellarAction>>,
    assistive_overrides: HashMap<KeyChord, StellarAction>,
}

impl KeymapEngine {
    pub fn new() -> Self {
        let mut engine = Self::default();
        engine.populate_global();
        engine.populate_persona_overlays();
        engine.populate_assistive();
        engine
    }

    fn populate_global(&mut self) {
        use StellarAction::*;
        self.global.insert(
            KeyChord::new(KeyCode::Tab, KeyModifiers::NONE),
            NavigateNextPane,
        );
        self.global.insert(
            KeyChord::new(KeyCode::BackTab, KeyModifiers::SHIFT),
            NavigatePrevPane,
        );
        self.global.insert(
            KeyChord::new(KeyCode::Char('k'), KeyModifiers::CONTROL),
            OpenCommandPalette,
        );
        self.global.insert(
            KeyChord::new(
                KeyCode::Char('c'),
                KeyModifiers::CONTROL | KeyModifiers::SHIFT,
            ),
            StellarAction::OpenConflictOverlay,
        );
        self.global.insert(
            KeyChord::new(KeyCode::Char('i'), KeyModifiers::CONTROL),
            ToggleCanvas,
        );
        self.global.insert(
            KeyChord::new(KeyCode::Char('o'), KeyModifiers::CONTROL),
            ToggleTelemetryOverlay,
        );
        self.global.insert(
            KeyChord::new(KeyCode::Char('r'), KeyModifiers::CONTROL),
            RunbookInvoke { runbook_id: None },
        );
        self.global.insert(
            KeyChord::new(KeyCode::Char('z'), KeyModifiers::CONTROL),
            InputUndo,
        );
        self.global.insert(
            KeyChord::new(
                KeyCode::Char('z'),
                KeyModifiers::CONTROL | KeyModifiers::SHIFT,
            ),
            InputRedo,
        );
        self.global.insert(
            KeyChord::new(KeyCode::Char('y'), KeyModifiers::CONTROL),
            InputRedo,
        );
        self.global.insert(
            KeyChord::new(KeyCode::Char('l'), KeyModifiers::CONTROL),
            FieldLockToggle,
        );
        self.global.insert(
            KeyChord::new(KeyCode::Char('i'), KeyModifiers::NONE),
            ToggleConfidencePanel,
        );
        self.global.insert(
            KeyChord::new(KeyCode::Enter, KeyModifiers::CONTROL),
            SubmitInsight { text: None },
        );
        self.global.insert(
            KeyChord::new(
                KeyCode::Char('a'),
                KeyModifiers::CONTROL | KeyModifiers::ALT,
            ),
            AccessibilityToggle,
        );
    }

    fn populate_persona_overlays(&mut self) {
        use StellarAction::*;
        let mut sre = HashMap::new();
        sre.insert(
            KeyChord::new(
                KeyCode::Char('o'),
                KeyModifiers::CONTROL | KeyModifiers::SHIFT,
            ),
            ToggleTelemetryOverlay,
        );
        self.persona_overlays.insert(StellarPersona::Sre, sre);

        let mut secops = HashMap::new();
        secops.insert(
            KeyChord::new(
                KeyCode::Char('k'),
                KeyModifiers::CONTROL | KeyModifiers::SHIFT,
            ),
            OpenCommandPalette,
        );
        self.persona_overlays.insert(StellarPersona::SecOps, secops);

        let mut platform = HashMap::new();
        platform.insert(
            KeyChord::new(KeyCode::Enter, KeyModifiers::ALT),
            SubmitInsight { text: None },
        );
        self.persona_overlays
            .insert(StellarPersona::PlatformEngineer, platform);
    }

    fn populate_assistive(&mut self) {
        use StellarAction::*;
        self.assistive_overrides.insert(
            KeyChord::new(KeyCode::Right, KeyModifiers::CONTROL),
            NavigateNextPane,
        );
        self.assistive_overrides.insert(
            KeyChord::new(KeyCode::Left, KeyModifiers::CONTROL),
            NavigatePrevPane,
        );
        self.assistive_overrides.insert(
            KeyChord::new(KeyCode::Char('i'), KeyModifiers::ALT),
            ToggleCanvas,
        );
        self.assistive_overrides.insert(
            KeyChord::new(KeyCode::F(1), KeyModifiers::NONE),
            OpenCommandPalette,
        );
        self.assistive_overrides.insert(
            KeyChord::new(KeyCode::F(6), KeyModifiers::NONE),
            ToggleTelemetryOverlay,
        );
        self.assistive_overrides.insert(
            KeyChord::new(KeyCode::F(7), KeyModifiers::NONE),
            StellarAction::OpenConflictOverlay,
        );
        self.assistive_overrides.insert(
            KeyChord::new(KeyCode::F(9), KeyModifiers::NONE),
            RunbookInvoke { runbook_id: None },
        );
        self.assistive_overrides.insert(
            KeyChord::new(KeyCode::Backspace, KeyModifiers::ALT),
            InputUndo,
        );
        self.assistive_overrides.insert(
            KeyChord::new(KeyCode::Enter, KeyModifiers::SHIFT),
            ToggleConfidencePanel,
        );
        self.assistive_overrides.insert(
            KeyChord::new(KeyCode::F(10), KeyModifiers::NONE),
            AccessibilityToggle,
        );
    }

    pub fn resolve(
        &self,
        persona: StellarPersona,
        assistive_mode: bool,
        event: KeyEvent,
    ) -> Option<StellarAction> {
        if !matches!(event.kind, KeyEventKind::Press | KeyEventKind::Repeat) {
            return None;
        }
        let chord = KeyChord::new(event.code, event.modifiers);
        if assistive_mode {
            if let Some(action) = self.assistive_overrides.get(&chord) {
                return Some(with_payload(action.clone(), event));
            }
        }
        if let Some(persona_map) = self.persona_overlays.get(&persona) {
            if let Some(action) = persona_map.get(&chord) {
                return Some(with_payload(action.clone(), event));
            }
        }
        self.global
            .get(&chord)
            .cloned()
            .map(|action| with_payload(action, event))
    }
}

fn with_payload(action: StellarAction, event: KeyEvent) -> StellarAction {
    match action {
        StellarAction::RunbookInvoke { runbook_id } => {
            if runbook_id.is_some() {
                return StellarAction::RunbookInvoke { runbook_id };
            }
            // Allow launching runbooks via numeric function keys, e.g. F2 -> RB-02.
            if let KeyCode::F(n) = event.code {
                let id = format!("RB-{n:02}");
                return StellarAction::RunbookInvoke {
                    runbook_id: Some(id),
                };
            }
            StellarAction::RunbookInvoke { runbook_id: None }
        }
        StellarAction::SubmitInsight { text: payload } => {
            if payload.is_some() {
                return StellarAction::SubmitInsight { text: payload };
            }
            // Allow typing text via modifier chords (Ctrl+Enter carries no text).
            StellarAction::SubmitInsight { text: None }
        }
        other => other,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crossterm::event::KeyCode;
    use crossterm::event::KeyEvent;
    use crossterm::event::KeyEventKind;
    use crossterm::event::KeyModifiers;

    #[test]
    fn ctrl_enter_maps_to_submit() {
        let engine = KeymapEngine::new();
        let event = KeyEvent {
            code: KeyCode::Enter,
            modifiers: KeyModifiers::CONTROL,
            kind: KeyEventKind::Press,
            state: crossterm::event::KeyEventState::NONE,
        };
        let action = engine
            .resolve(StellarPersona::Operator, false, event)
            .expect("should map to submit");
        match action {
            StellarAction::SubmitInsight { text } => assert!(text.is_none()),
            other => panic!("expected submit, got {other:?}"),
        }
    }

    #[test]
    fn ctrl_shift_c_opens_conflict_overlay() {
        let engine = KeymapEngine::new();
        let event = KeyEvent {
            code: KeyCode::Char('c'),
            modifiers: KeyModifiers::CONTROL | KeyModifiers::SHIFT,
            kind: KeyEventKind::Press,
            state: crossterm::event::KeyEventState::NONE,
        };
        let action = engine
            .resolve(StellarPersona::Operator, false, event)
            .expect("should map to conflict overlay");
        assert!(matches!(action, StellarAction::OpenConflictOverlay));
    }
}
