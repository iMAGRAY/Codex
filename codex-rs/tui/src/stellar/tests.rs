use super::StellarController;
use super::StellarView;
use codex_core::stellar::StellarPersona;
use crossterm::event::KeyCode;
use crossterm::event::KeyEvent;
use crossterm::event::KeyModifiers;
use insta::assert_snapshot;
use ratatui::Terminal;
use ratatui::backend::TestBackend;

fn send(ctrl: &mut StellarController, code: KeyCode, modifiers: KeyModifiers) {
    let event = KeyEvent::new(code, modifiers);
    let _ = ctrl.handle_key_event(event);
}

fn activate_canvas(ctrl: &mut StellarController) {
    send(ctrl, KeyCode::Char('i'), KeyModifiers::CONTROL);
}

fn seed_insight(ctrl: &mut StellarController) {
    ctrl.set_field_text("Investigate APDEX drift");
    send(ctrl, KeyCode::Char('o'), KeyModifiers::CONTROL);
    send(ctrl, KeyCode::Char('i'), KeyModifiers::NONE);
    send(ctrl, KeyCode::Enter, KeyModifiers::CONTROL);
}

#[test]
fn stellar_view_wide_snapshot() {
    let mut controller = StellarController::new(StellarPersona::Operator);
    activate_canvas(&mut controller);
    seed_insight(&mut controller);
    controller.sync_layout(140);
    controller.take_events();

    let snapshot = controller.snapshot();
    let mut terminal = Terminal::new(TestBackend::new(140, 20)).expect("terminal");
    terminal
        .draw(|f| {
            f.render_widget(StellarView::new(&snapshot), f.area());
        })
        .expect("draw");

    assert_snapshot!("stellar_view_wide", terminal.backend());
}

#[test]
fn stellar_view_compact_snapshot() {
    let mut controller = StellarController::new(StellarPersona::Operator);
    activate_canvas(&mut controller);
    seed_insight(&mut controller);
    controller.sync_layout(90);
    controller.take_events();

    let snapshot = controller.snapshot();
    let mut terminal = Terminal::new(TestBackend::new(90, 22)).expect("terminal");
    terminal
        .draw(|f| {
            f.render_widget(StellarView::new(&snapshot), f.area());
        })
        .expect("draw");

    assert_snapshot!("stellar_view_compact", terminal.backend());
}

#[test]
fn ctrl_o_toggles_telemetry_overlay() {
    let mut controller = StellarController::new(StellarPersona::Operator);
    activate_canvas(&mut controller);

    assert!(!controller.snapshot().telemetry_visible);

    send(&mut controller, KeyCode::Char('o'), KeyModifiers::CONTROL);
    assert!(controller.snapshot().telemetry_visible);

    send(&mut controller, KeyCode::Char('o'), KeyModifiers::CONTROL);
    assert!(!controller.snapshot().telemetry_visible);
}
