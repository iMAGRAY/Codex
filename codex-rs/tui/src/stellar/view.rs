use crate::status_bar::telemetry_spans;
use codex_core::stellar::GoldenPathHint;
use codex_core::stellar::KernelSnapshot;
use codex_core::stellar::LayoutMode;
use codex_core::stellar::PaneFocus;
use codex_core::stellar::RiskSeverity;
use codex_core::stellar::StellarPersona;
use codex_core::telemetry::TelemetryHub;
use ratatui::buffer::Buffer;
use ratatui::layout::Constraint;
use ratatui::layout::Direction;
use ratatui::layout::Layout;
use ratatui::layout::Rect;
use ratatui::style::Color;
use ratatui::style::Modifier;
use ratatui::style::Style;
use ratatui::style::Stylize;
use ratatui::text::Line;
use ratatui::text::Span;
use ratatui::widgets::Block;
use ratatui::widgets::Borders;
use ratatui::widgets::Paragraph;
use ratatui::widgets::Widget;
use ratatui::widgets::WidgetRef;
use ratatui::widgets::Wrap;

pub struct StellarView<'a> {
    snapshot: &'a KernelSnapshot,
}

impl<'a> StellarView<'a> {
    pub fn new(snapshot: &'a KernelSnapshot) -> Self {
        Self { snapshot }
    }

    fn render_wide(&self, area: Rect, buf: &mut Buffer) {
        if area.height <= 3 || area.width <= 20 {
            return;
        }
        let vertical = Layout::default()
            .direction(Direction::Vertical)
            .constraints([
                Constraint::Min(area.height.saturating_sub(4)),
                Constraint::Length(3),
            ])
            .split(area);

        let main = Layout::default()
            .direction(Direction::Horizontal)
            .constraints([Constraint::Percentage(70), Constraint::Percentage(30)])
            .split(vertical[0]);

        self.render_canvas(main[0], buf);
        self.render_telemetry(main[1], buf);

        let footer = Layout::default()
            .direction(Direction::Horizontal)
            .constraints([Constraint::Percentage(60), Constraint::Percentage(40)])
            .split(vertical[1]);

        self.render_command_log(footer[0], buf);
        self.render_golden_path(footer[1], buf);
    }

    fn render_compact(&self, area: Rect, buf: &mut Buffer) {
        if area.height <= 6 {
            return;
        }
        let layout = Layout::default()
            .direction(Direction::Vertical)
            .constraints([
                Constraint::Min(area.height.saturating_sub(6)),
                Constraint::Length(2),
                Constraint::Length(2),
                Constraint::Length(2),
            ])
            .split(area);

        self.render_canvas(layout[0], buf);
        self.render_tab_strip(layout[1], buf);
        self.render_command_log(layout[2], buf);
        self.render_golden_path(layout[3], buf);
    }

    fn render_canvas(&self, area: Rect, buf: &mut Buffer) {
        let focus = self.snapshot.focus == PaneFocus::InsightCanvas;
        let mut block = styled_block("Insight Canvas", focus);
        if self.snapshot.field_locked {
            block = block.title(" Insight Canvas (locked) ");
        }
        block.render_ref(area, buf);
        if area.height <= 2 {
            return;
        }
        let inner = block.inner(area);
        let chunks = Layout::default()
            .direction(Direction::Vertical)
            .constraints([
                Constraint::Length(3),
                Constraint::Length(3),
                Constraint::Min(inner.height.saturating_sub(6)),
            ])
            .split(inner);

        let persona = format!(
            "Persona: {}{}",
            self.snapshot.persona,
            if self.snapshot.assistive_mode {
                " (assistive)"
            } else {
                ""
            }
        );
        let status_line = Paragraph::new(Line::from(persona.cyan()));
        status_line.render_ref(chunks[0], buf);

        let input = if self.snapshot.field_text.is_empty() {
            "<type an insight or use CLI bridge>".italic()
        } else {
            self.snapshot.field_text.clone().into()
        };
        Paragraph::new(input)
            .wrap(Wrap { trim: true })
            .render_ref(chunks[1], buf);

        self.render_suggestions(chunks[2], buf);
    }

    fn render_suggestions(&self, area: Rect, buf: &mut Buffer) {
        let mut lines = Vec::new();
        if !self.snapshot.risk_alerts.is_empty() {
            for alert in &self.snapshot.risk_alerts {
                let styled = match alert.severity {
                    RiskSeverity::Info => format!("ℹ {}", alert.message).blue(),
                    RiskSeverity::Warning => format!("⚠ {}", alert.message).yellow().bold(),
                    RiskSeverity::Critical => format!("⛔ {}", alert.message).red().bold(),
                };
                lines.push(Line::from(styled));
            }
            lines.push(Line::from(Span::raw("")));
        }
        if let Some(confidence) = &self.snapshot.confidence {
            if confidence.visible {
                lines.push(Line::from(vec![
                    Span::from(format!(
                        "Confidence {:>3}% - {}",
                        (confidence.score * 100.0).round() as i32,
                        confidence.trend
                    ))
                    .green(),
                ]));
                for reason in &confidence.reasons {
                    lines.push(prefix_bullet(reason));
                }
                lines.push(Line::from(Span::raw("")));
            }
        }
        for suggestion in &self.snapshot.suggestions {
            lines.push(prefix_bullet(suggestion));
        }
        if self.snapshot.status_messages.is_empty() {
            lines.push(Line::from("Press Ctrl+Enter to submit".dim()));
        } else {
            for status in &self.snapshot.status_messages {
                lines.push(Line::from(status.clone().yellow()));
            }
        }
        Paragraph::new(lines)
            .wrap(Wrap { trim: true })
            .render_ref(area, buf);
    }

    fn render_telemetry(&self, area: Rect, buf: &mut Buffer) {
        let focus = self.snapshot.focus == PaneFocus::Telemetry;
        let title = if self.snapshot.telemetry_visible {
            "Observability Overlay"
        } else {
            "Observability Overlay (hidden)"
        };
        let block = styled_block(title, focus);
        block.render_ref(area, buf);
        if area.height <= 2 {
            return;
        }
        let inner = block.inner(area);
        if self.snapshot.telemetry_visible {
            let snapshot = TelemetryHub::global().snapshot();
            let mut lines = Vec::new();
            lines.push(Line::from(telemetry_spans(snapshot)));
            lines.push(investigate_hint(
                self.snapshot.persona,
                self.snapshot.assistive_mode,
            ));
            lines.push(Line::from("Ctrl+O to hide overlay".cyan()));
            Paragraph::new(lines)
                .wrap(Wrap { trim: true })
                .render_ref(inner, buf);
        } else {
            let mut lines = Vec::new();
            lines.push(Line::from("Press Ctrl+O to reveal live metrics".cyan()));
            lines.push(investigate_hint(
                self.snapshot.persona,
                self.snapshot.assistive_mode,
            ));
            Paragraph::new(lines)
                .wrap(Wrap { trim: true })
                .render_ref(inner, buf);
        }
    }

    fn render_command_log(&self, area: Rect, buf: &mut Buffer) {
        let focus = self.snapshot.focus == PaneFocus::CommandLog;
        let block = styled_block("Command Log", focus);
        block.render_ref(area, buf);
        if area.height == 0 {
            return;
        }
        let inner = block.inner(area);
        let lines: Vec<_> = if self.snapshot.command_log.is_empty() {
            vec![Line::from("No insights submitted yet".dim())]
        } else {
            self.snapshot
                .command_log
                .iter()
                .rev()
                .take(inner.height as usize)
                .map(|entry| Line::from(entry.clone()))
                .collect()
        };
        Paragraph::new(lines).render_ref(inner, buf);
    }

    fn render_golden_path(&self, area: Rect, buf: &mut Buffer) {
        let focus = matches!(
            self.snapshot.focus,
            PaneFocus::FooterSubmit | PaneFocus::FooterOverlay
        );
        let block = styled_block("Golden Path", focus);
        block.render_ref(area, buf);
        let inner = block.inner(area);
        let mut lines = Vec::new();
        for hint in &self.snapshot.golden_path {
            lines.push(render_hint(hint));
        }
        Paragraph::new(lines)
            .wrap(Wrap { trim: true })
            .render_ref(inner, buf);
    }

    fn render_tab_strip(&self, area: Rect, buf: &mut Buffer) {
        let tabs = [
            ("Telemetry", PaneFocus::Telemetry),
            ("Command Log", PaneFocus::CommandLog),
            ("Runbook", PaneFocus::Runbook),
        ];
        let spans: Vec<Span> = tabs
            .into_iter()
            .map(|(label, focus)| {
                if self.snapshot.focus == focus {
                    Span::from(format!("[{label}] ")).bold()
                } else {
                    Span::from(format!(" {label} ")).dim()
                }
            })
            .collect();
        Paragraph::new(Line::from(spans)).render_ref(area, buf);
    }
}

impl<'a> Widget for StellarView<'a> {
    fn render(self, area: Rect, buf: &mut Buffer) {
        match self.snapshot.layout_mode {
            LayoutMode::Wide => self.render_wide(area, buf),
            LayoutMode::Compact => self.render_compact(area, buf),
        }
    }
}

fn investigate_hint(persona: StellarPersona, assistive_mode: bool) -> Line<'static> {
    let key = if assistive_mode { "F9" } else { "Ctrl+R" };
    let detail = match persona {
        StellarPersona::Operator => "guide the operator runbook",
        StellarPersona::Sre => "run SRE latency playbook RB-07",
        StellarPersona::SecOps => "review audit and policy evidence",
        StellarPersona::PlatformEngineer => "inspect deployment timeline",
        StellarPersona::PartnerDeveloper => "handoff to partner support flow",
        StellarPersona::AssistiveBridge => "narrate investigation with accessible summary",
    };
    Line::from(vec![
        Span::from("[ Investigate ]").cyan().bold(),
        Span::raw(" "),
        Span::from(format!("{key} - {detail}")),
    ])
}

fn styled_block<'a>(title: &'a str, focused: bool) -> Block<'a> {
    let style = if focused {
        Style::default()
            .fg(Color::Cyan)
            .add_modifier(Modifier::BOLD)
    } else {
        Style::default()
    };
    Block::default()
        .title(Span::styled(format!(" {title} "), style))
        .borders(Borders::ALL)
        .border_style(style)
}

fn prefix_bullet(text: &str) -> Line<'_> {
    Line::from(vec![" - ".into(), Span::raw(text.to_string())])
}

fn render_hint(hint: &GoldenPathHint) -> Line<'_> {
    let mut spans = Vec::new();
    spans.push(Span::raw(hint.label.clone()).bold());
    if let Some(shortcut) = &hint.shortcut {
        spans.push(Span::raw(" "));
        spans.push(Span::from(shortcut.clone()).cyan());
    }
    spans.push(Span::raw(" — "));
    spans.push(Span::raw(hint.description.clone()));
    Line::from(spans)
}
