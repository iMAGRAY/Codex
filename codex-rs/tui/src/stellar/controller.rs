use crate::stellar::keymap::KeymapEngine;
use codex_core::resilience::CacheConfig;
use codex_core::resilience::QueueConfig;
use codex_core::resilience::ResilienceCache;
use codex_core::resilience::ResilienceServices;
use codex_core::resilience::RetryQueue;
use codex_core::stellar::ActionApplied;
use codex_core::stellar::GuardError;
use codex_core::stellar::KernelEvent;
use codex_core::stellar::KernelSnapshot;
use codex_core::stellar::LayoutMode;
use codex_core::stellar::StellarAction;
use codex_core::stellar::StellarCliEvent;
use codex_core::stellar::StellarKernel;
use codex_core::stellar::StellarPersona;
use crossterm::event::KeyEvent;

#[derive(Debug)]
pub struct StellarController {
    kernel: StellarKernel,
    keymap: KeymapEngine,
    resilience: ResilienceServices,
}

#[derive(Debug, Clone)]
pub enum ControllerOutcome {
    Consumed {
        #[allow(dead_code)]
        action: StellarAction,
        applied: ActionApplied,
    },
    Rejected {
        #[allow(dead_code)]
        action: StellarAction,
        error: GuardError,
    },
    Unhandled,
}

impl StellarController {
    pub fn new(persona: StellarPersona) -> Self {
        let resilience =
            ResilienceServices::default().unwrap_or_else(|_| fallback_resilience_services());
        let kernel = StellarKernel::with_resilience(persona, resilience.clone());
        Self {
            kernel,
            keymap: KeymapEngine::new(),
            resilience,
        }
    }

    pub fn persona(&self) -> StellarPersona {
        self.kernel.persona()
    }

    pub fn is_active(&self) -> bool {
        self.kernel.is_canvas_visible()
    }

    pub fn assistive_mode(&self) -> bool {
        self.kernel.assistive_mode()
    }

    pub fn handle_key_event(&mut self, event: KeyEvent) -> ControllerOutcome {
        let persona = self.persona();
        let assistive = self.assistive_mode();
        let Some(action) = self.keymap.resolve(persona, assistive, event) else {
            return ControllerOutcome::Unhandled;
        };
        match self.kernel.handle_action(action.clone()) {
            Ok(applied) => ControllerOutcome::Consumed { action, applied },
            Err(error) => ControllerOutcome::Rejected { action, error },
        }
    }

    #[allow(dead_code)]
    pub fn apply_cli_event(
        &mut self,
        cli_event: StellarCliEvent,
    ) -> Result<ActionApplied, GuardError> {
        if cli_event.persona != self.persona() {
            self.kernel =
                StellarKernel::with_resilience(cli_event.persona, self.resilience.clone());
        }
        self.kernel.handle_action(cli_event.action)
    }

    pub fn snapshot(&self) -> KernelSnapshot {
        self.kernel.snapshot()
    }

    pub fn take_events(&mut self) -> Vec<KernelEvent> {
        self.kernel.take_events()
    }

    pub fn preferred_height(&self) -> u16 {
        match self.kernel.layout_mode() {
            LayoutMode::Wide => 18,
            LayoutMode::Compact => 22,
        }
    }

    pub fn sync_layout(&mut self, width: u16) -> LayoutMode {
        let mode = if width >= 120 {
            LayoutMode::Wide
        } else {
            LayoutMode::Compact
        };
        self.kernel.set_layout_mode(mode);
        mode
    }

    #[allow(dead_code)]
    pub fn set_field_text(&mut self, value: impl Into<String>) {
        self.kernel.set_field_text(value);
    }

    #[allow(dead_code)]
    pub fn clear_status(&mut self) {
        self.kernel.clear_status_messages();
    }
}

fn fallback_resilience_services() -> ResilienceServices {
    let cache = ResilienceCache::open(CacheConfig::default()).expect("fallback cache init failed");
    let queue = RetryQueue::open(QueueConfig::default()).expect("fallback queue init failed");
    ResilienceServices::new(cache, queue)
}
