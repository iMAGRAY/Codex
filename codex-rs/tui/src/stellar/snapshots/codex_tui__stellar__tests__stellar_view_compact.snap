---
source: tui/src/stellar/tests.rs
assertion_line: 62
expression: terminal.backend()
---
"┌ Insight Canvas ────────────────────────────────────────────────────────────────────────┐"
"│Persona: operator                                                                       │"
"│                                                                                        │"
"│                                                                                        │"
"│<type an insight or use CLI bridge>                                                     │"
"│                                                                                        │"
"│                                                                                        │"
"│⚠ 1 queued offline operation                                                            │"
"│                                                                                        │"
"│Confidence  94% - High                                                                  │"
"│- Freshness: 0.35                                                                       │"
"│- Source Trust: 0.27                                                                    │"
"│- Schema Validity: 0.19                                                                 │"
"│- Telemetry Alignment: 0.10                                                             │"
"│- User Overrides: 0.03                                                                  │"
"└────────────────────────────────────────────────────────────────────────────────────────┘"
"[Telemetry]  Command Log  Runbook                                                         "
"                                                                                          "
"┌ Command Log ───────────────────────────────────────────────────────────────────────────┐"
"└────────────────────────────────────────────────────────────────────────────────────────┘"
"┌ Golden Path ───────────────────────────────────────────────────────────────────────────┐"
"└────────────────────────────────────────────────────────────────────────────────────────┘"
